File: spark-java/src/main/java/adj/spark/core/transformation/JoinOperation.java
package adj.spark.core.transformation;

import java.util.Arrays;
import java.util.List;

import org.apache.spark.SparkConf;
import org.apache.spark.api.java.JavaPairRDD;
import org.apache.spark.api.java.JavaSparkContext;
import org.apache.spark.api.java.function.VoidFunction;

import scala.Tuple2;

@SuppressWarnings("resource")
public class JoinOperation {
	public static void main(String[] args) {
		SparkConf conf = new SparkConf();
		          conf.setAppName("GroupByOperation")
		              .setMaster("local");
		JavaSparkContext jsc = new JavaSparkContext(conf);

		List<Tuple2<Integer, String>> studentList = Arrays.asList(
				new Tuple2<Integer, String>(1, "jack"),
				new Tuple2<Integer, String>(1, "mary"),
				new Tuple2<Integer, String>(2, "jack")
		);
		List<Tuple2<Integer, Integer>> scoreList = Arrays.asList(
				new Tuple2<Integer, Integer>(1, 10101),
				new Tuple2<Integer, Integer>(2, 10102),
				new Tuple2<Integer, Integer>(2, 10103)
		);
		
		// 并行化2个集合
		JavaPairRDD<Integer, String> studentRDD = jsc.parallelizePairs(studentList);
		JavaPairRDD<Integer, Integer> scoreRDD = jsc.parallelizePairs(scoreList);
		
		/**
		 * colgroup: key对应的值已经分组
		 * key -> ((v1_0,v1_2), (v2_0,v2_1))
		 * 
		 * join    : key对应的值没有分组
		 * key  -> (v1_0, v2_0)
		 * key  -> (v1_1, v2_1)
		 */
		JavaPairRDD<Integer, Tuple2<String, Integer>> joinRDD = studentRDD.join(scoreRDD);
		joinRDD.foreach(new VoidFunction<Tuple2<Integer,Tuple2<String,Integer>>>() {
			private static final long serialVersionUID = 1L;

			@Override
			public void call(Tuple2<Integer, Tuple2<String, Integer>> info) throws Exception {
				System.out.println("================");
				System.out.println(info._1);
				
				System.out.println(info._2._1 + " = " + info._2._2);
			}
		});
	}
}

File: spark-scala-maven/src/main/scala/adj/spark/core/transformation/JoinOperation.scala
package adj.spark.core.transformation

import org.apache.spark.SparkConf
import org.apache.spark.SparkContext

object JoinOperation {
  def main(args: Array[String]): Unit = {
    val conf = new SparkConf()
               .setMaster("local")
               .setAppName("FlatMapOperation")
        
    val sc = new SparkContext(conf)
        
    val users = Array((1, "jack"), (2, "kafka"), (3, "tom"), (4, "leo"), (4, "foko"))
    val roles = Array((1, 1), (2, 3), (3, 3), (4, 4), (4, 41))
    
    val userRDD = sc.parallelize(users, 3)
    val roleRDD = sc.parallelize(roles, 3)
    
    val colgroupRDD = userRDD.cogroup(roleRDD)
    colgroupRDD.foreach(g => {
      println(g._1)
      println(g._2._1)
      println(g._2._2)
    })
  }
}